### PR TITLE
Avoid infinite loop in case of heap corruption

### LINKED ISSUE #70750: DAC heapwalker doesn't detect memory corruption that makes objects appear to have zero size
### Description

When enumerating the GC heap with ICorDebugHeapEnum::Next the enumeration could get stuck repeatedly enumerating an object with size = 0. Because the size is zero internally the enumerator isn't advancing the pointer into heap which causes the repetition. We believe this occurs because the heap memory is corrupt, but regardless a debugging enumerator should not enter an infinite loop in the presence of bad data in a debuggee.

The likely cause is that the DAC never explicitly checks to see if the object size computed from the MethodTable is in a reasonable range:
[runtime/dacdbiimpl.cpp at 7015fa2d28e4559478807a53b5851c71d2f604e8 · dotnet/runtime (github.com)](https://github.com/dotnet/runtime/blob/7015fa2d28e4559478807a53b5851c71d2f604e8/src/coreclr/debug/daccess/dacdbiimpl.cpp#L6451)

```
bool DacHeapWalker::GetSize(TADDR tMT, size_t &size)
{
    // With heap corruption, it's entierly possible that the MethodTable
    // we get is bad.  This could cause exceptions, which we will catch
    // and return false.  This causes the heapwalker to move to the next
    // segment.
```

We should add such a check to this function.

### Reproduction Steps

There is a heap dump on an internal email that can be loaded in a debugger. I am also hypothesizing that any valid heap dump could be made to repro this issue by overwriting the size field in the MethodTable with a zero.

### Expected behavior

Enumerator skips past some range of bad objects (potentially all remaining objects) once the heap corruption is detected.

### Actual behavior

Enumerator goes into an infinite loop

### Regression?

Doesn't look like it

### Known Workarounds

The caller can detect the object size of zero and stop invoking Next() any further.

### Configuration

.NET 4.8 is where the problem reproed. Looking at the code the bug is still present in current versions of .NET, any arch or OS.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by cshung**
Fixes https://github.com/dotnet/runtime/issues/70750

### CODE REVIEW

File: src/coreclr/debug/daccess/dacdbiimpl.cpp [hunk 1]
```diff
@@ -6478,6 +6478,10 @@ bool DacHeapWalker::GetSize(TADDR tMT, size_t &size)
             size = AlignLarge(size);
         else
             size = Align(size);
+
+        // If size == 0, it means we have a heap corruption and
+        // we will stuck in an infinite loop, so better fail the call now.
+        ret = size > 0;
```
Comment (hoyosjs) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:6484]: Looks good as is, although would an upper bound check also make sense? Does the GC have an upper bound cap we could check here? I expect those to be more common corruptions than 0, although like it won't cause something as bad as the infinite look in the walk.

**Review by hoyosjs — APPROVED** [review]

File: src/coreclr/debug/daccess/dacdbiimpl.cpp (hunk 1 above)
Comment (cshung) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:6484]: We could upper bound it by the high alloc mark.

Comment (danmoseley) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:6484]: if line 6469 set ret to false, is it possible this could change it to true?

Comment (hoyosjs) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:6484]:
```suggestion
        ret |= size > 0;
```

**Review by tommcdon — APPROVED** [review]
Thanks Andrew!

File: src/coreclr/debug/daccess/dacdbiimpl.cpp [hunk 2]
```diff
@@ -6478,6 +6478,10 @@ bool DacHeapWalker::GetSize(TADDR tMT, size_t &size)
             size = AlignLarge(size);
         else
             size = Align(size);
+
+        // If size == 0, it means we have a heap corruption and
+        // we will stuck in an infinite loop, so better fail the call now.
+        ret |= ((0 < size) && (tMT + size) <= mHeaps[mCurrHeap].Segments[mCurrSeg].End);
```
[the remark this answers is not in the document]
Comment (hoyosjs) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:6484]: Oops :)

File: src/coreclr/debug/daccess/dacdbiimpl.cpp [hunk 3]
```diff
@@ -6478,6 +6478,11 @@ bool DacHeapWalker::GetSize(TADDR tMT, size_t &size)
             size = AlignLarge(size);
         else
             size = Align(size);
+
+        // If size == 0, it means we have a heap corruption and
+        // we will stuck in an infinite loop, so better fail the call now.
+        ret &= (0 < size);
+        ret &= ((tMT + size) <= mHeaps[mCurrHeap].Segments[mCurrSeg].End);
```
Comment (hoyosjs) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:6485]:
```suggestion
        // Also guard for cases where the size is large and exceeds the high allocation mark.
        ret &= ((tMT + size) <= mHeaps[mCurrHeap].Segments[mCurrSeg].End);
```

**Pull request merged**